infrastructure-userdb/display-name.rkt

```diff
@@ -0,0 +1,146 @@
+;;     needed. We have no new data fields to store or validate.
+;;
+;;  3. There is no ambiguity between email addresses and obfuscated display
+;;     names: the former always contain `@`; the later never do.
+;;
+;;  4. Anyone who knows an email address can compute the corresponding
+;;     obfuscated display name and thus can search for packages associated
+;;     with it. In some contexts one would want to use a UUID or a salted hash
+;;     to avoid making that information discoverable. Here, though, our goal
+;;     is to protect package authors from being spammed, not to conceal the
+;;     authorship of Racket packages.
+;;
+;; All of this module’s exports are pure functions. This module doesn’t depend
+;; on any other parts of this package, but it provides sufficient primitives
+;; to implement an obfuscation opt-out mechanism elsewhere using user properties.
+;;
+
+(require racket/match
+         racket/contract
+         (only-in file/sha1 bytes->hex-string)
+         (only-in xml xexpr/c))
+
+(provide display-name?
+         (contract-out
+          [email->display-name
+           (-> string? display-name?)]
+          [display-name-email
+           (-> display-name? string?)]
+          [display-name-obfuscated-email
+           (-> display-name? string?)]
+          [display-name-local-part
+           (-> display-name? string?)]
+          [display-name-short-hash
+           (-> display-name? string?)]
+          [display-name-tags
+           (-> display-name? (listof symbol?))]
+          [display-name-plain-tag
+           (-> display-name? symbol?)]
+          [display-name-obfuscated-tag
+           (-> display-name? symbol?)]))
+
+(module+ test
+  (require (submod ".."))) ; enforce contracts
+
+(struct display-name (email local-part short-hash plain-tag obfuscated-tag)
+  #:guard (struct-guard/c string? string? string? symbol? symbol?)
+  #:transparent)
+
+(define email->display-name
+  (let ([plain-tag-memo (make-weak-hasheq)]
+        [display-name-memo (make-ephemeron-hasheq)])
+    ;; indirection because equal?-based hashes aren't thread-safe
+    (define (email->display-name email)
+      (define plain-tag
+        (hash-ref! plain-tag-memo
+                   email
+                   (λ ()
+                     (email->plain-tag email))))
+      (hash-ref! display-name-memo
+                 plain-tag
+                 (λ ()
+                   (make-display-name email plain-tag))))
+    email->display-name))
+
+(define (email->plain-tag email)
+  (string->symbol (string-append "author:" email)))
+
+(define (make-display-name email [plain-tag (email->plain-tag email)])
+  (match-define (list local-part)
+    (regexp-match #rx"^[^@]*" email))
+  (define short-hash
+    (substring (bytes->hex-string
+                (sha256-bytes (open-input-string email)))
+               0
+               7)) ; on the very scientific basis that this is what Git uses
```

Comment:
That is true, but:
1. Personally, I've never seen an example of a Git repository that has needed more that 7 digits to avoid a collision.
2. In our case, there would only be confusion if the colliding email addresses also have the same "local part".

Reply:
> 1. Personally, I've never seen an example of a Git repository that has needed more that 7 digits to avoid a collision.

They're not that hard to find if you look at massive projects... For example, the LLVM repo I have on hand currently needs 12 digits for its 450k commits. Mozilla's gecko-dev repo (Firefox) needs 13 digits for 800k commits.

In any case, as you say, it's unlikely to be an issue here.